exprel/graph_extractor/extract.py

```diff
@@ -0,0 +1,331 @@
+                        "-", "_"
+                    )
+                node_lines.append(node_line)
+            lines += sorted(node_lines)
+
+            edge_lines = []
+            for u, v, edata in graph.edges(data=True):
+                if "color" in edata:
+                    edge_lines.append(
+                        '\t{0} -> {1} [ label = "{2}" ];'.format(u, v, edata["color"])
+                    )
+
+            lines += sorted(edge_lines)
+            lines.append("}")
+            return "\n".join(lines)
+
+        graphs = {}
+        if os.path.isfile("longman_zero_paths_one_exp"):
+            with open("longman_zero_paths_one_exp.json") as f:
+                graphs = json.load(f)
+
+        words = {}
+        for fields in trained_features:
+            words[fields[1] + "_" + fields[3]] = int(fields[3])
+            feature = fields[0]
+        graph = nx.MultiDiGraph()
+
+        color_map = []
+        for word in words:
+            if words[word] == 1:
+                color = "green"
+            else:
+                color = "red"
+            graph.add_node(word, color=color)
+            word_clean = word.split("_")[0]
+            if word_clean in graphs:
+                hypernyms = graphs[word_clean]
+                for hypernym in hypernyms:
+                    hypernym_words = hypernyms[hypernym]
+                    for w in hypernym_words:
+                        if hypernym == "1":
+                            graph.add_edge(word, w, color=hypernym)
+
+        d = Source(to_dot(graph, feature))
+        d.engine = "circo"
+        d.format = "png"
+
+        selected_words = self.select_words(trained_features)
+
+        word_features = []
+
+        word_features.append(feature.replace(".*", "|".join(selected_words)))
+
+        return d.render(view=True), word_features
+
+    def select_words(self, trained_features):
+        features = []
+        labels = []
+
+        for fields in trained_features:
+            features.append(fields[1])
+            labels.append(int(fields[3]))
+        words_to_measures = {
+            word: {"TP": 0, "FP": 0, "TN": 0, "FN": 0} for word in set(features)
+        }
+        for word in words_to_measures:
+            for i, label in enumerate(labels):
+                if label and features[i] == word:
+                    words_to_measures[word]["TP"] += 1
+                if label and features[i] != word:
+                    words_to_measures[word]["FN"] += 1
+                if not label and features[i] == word:
+                    words_to_measures[word]["FP"] += 1
+                if not label and features[i] != word:
+                    words_to_measures[word]["TN"] += 1
+
+        for word in words_to_measures:
+            TP = words_to_measures[word]["TP"]
+            FP = words_to_measures[word]["FP"]
+            TN = words_to_measures[word]["TN"]
+            FN = words_to_measures[word]["FN"]
+
+            precision = TP / (TP + FP)
```

Comment:
consider using [`count_p_r_f`](https://github.com/recski/tuw-nlp/blob/db3277dfadf5ec923818dbd33846af3bfdde2c34/tuw_nlp/common/eval.py#L33)

Reply:
Thanks, will do it!